twilio_integration/twilio_integration/doctype/whatsapp_campaign/whatsapp_campaign.py

```diff
@@ -0,0 +1,86 @@
+# Copyright (c) 2021, Frappe and contributors
+# For license information, please see license.txt
+
+import frappe
+from frappe.model.document import Document
+from frappe.utils import get_site_url
+from twilio_integration.twilio_integration.doctype.whatsapp_message.whatsapp_message import WhatsAppMessage
+
+class WhatsAppCampaign(Document):
+	def validate(self):
+		if self.scheduled_time and self.status != 'Completed':
+			current_time = frappe.utils.now_datetime()
+			scheduled_time = frappe.utils.get_datetime(self.scheduled_time)
+
+			if scheduled_time < current_time:
+				frappe.throw(_("Scheduled Time must be a future time."))
+
+			self.status = 'Scheduled'
+
+		self.all_missing_recipients()
+	
+	def validate_attachment(self):
+		attachment = self.get_attachment()
+		if attachment:
+			if attachment.file_size > 16777216:
+				frappe.throw(_('Attachment size must be less than 16MB.'))
+
+			if attachment.is_private:
+				frappe.throw(_('Attachment must be public.'))
+
+			if attachment.get_extension() not in ['jpg', 'jpeg', 'png', 'mp3', 'ogg', 'amr', 'pdf', 'mp4']:
+				frappe.throw(_('Attachment format not supported.'))
+
+	def get_attachment(self):
+		file = frappe.db.get_value("File", {"attached_to_name": self.doctype, "attached_to_doctype": self.name, "is_private":0}, 'name')
+
+		if file:
+			return frappe.get_doc('File', file)
+		return None
+
+	def get_whatsapp_contact(self):
+		contacts = []
+
+		for row in self.recipients:
+			if row.whatsapp_no:
+				contacts.append(row.whatsapp_no)	
+		
+		return contacts
```

Comment:
[recipient.whatsapp_no for recipient in self.recipients if recipient.whatsapp_no]